fs/fs.go

```diff
@@ -932,6 +940,13 @@ func (ft *Features) WrapsFs(f Fs, w Fs) *Features {
 	return ft
 }
 
+// Resumer is an optional interface for Fs
+type Resumer interface {
+	// Resume checks whether the (remote, ID) pair is valid and returns
+	// the point the file should be resumed from or an error.
+	Resume(ctx context.Context, remote, ID, hashName, hashState string) (Pos int64, err error)
```

Comment:
I don't think I agree with the idea that a hash should be generally provided by backends. That means that all backends must be able to provide a hash.

Reply:
No. The proper supports that chunker feature but other backends are not required to support it.

from https://github.com/rclone/rclone/pull/4547#issuecomment-786889739
> if backend does not support it, its interface will return a nil accrued hashsum

Reply:
but the hash is in the interface signature

Reply:
It's in signature but it can be nil.

Reply:
> resumable uploads via a different mechanism.

This interface is generic enough but optional. Backends are free to ignore it. Please take a closer look at the source code.

Reply:
Here is quick synopsis of the code.
There are three parts here: the backend, the resume proper which keeps local cache of partial uploads and generic interface between them.
The flow is:
- During upload the resume proper takes snapshot by requesting supported info from backend.
- If later user decides to attempt resuming, the proper performs cache lookup and queries backend providing available info and a callback.
- The backend picks up supported pieces from the provided info, matches it with upstream, chooses resume identifier and calks back to the proper
- The resume aborts or continues

Notes: Rolling hashsums are just part of matched info along with size or name. They are generally optional. They can be supported by local or chunker backends. They provide additional data safety against broken uploads. There must be a good reason to drop them.

Reply:
> resume proper which keeps local cache of partial uploads

Do you mean the cache is not stored on the remote? You can't assume that there is a local filesystem available to rclone.

> resume proper takes snapshot by requesting supported info from backend

rclone should make effort to support resumable uploads regardless of if the backend supports it or not. This can be done with a configurable fallback backend for intermediate files

> There must be a good reason to drop them.

I am not proposing that we drop the hashes, they just need to be a separate interface, complimentary to the resumer interface. I quite like the idea of the rolling hashsum, the implementation should just be composable rather than ignorable parameters.

Reply:
> Do you mean the cache is not stored on the remote?

It's stored locally under ~/.cache/rclone/uploads or whatever

> You can't assume that there is a local filesystem available to rclone.

Indeed rclone keeps configs and vfs cache locally.

> rclone should make effort to support resumable uploads regardless of if the backend supports it or not. This can be done with a configurable fallback backend for intermediate files

Supporting "fallback upload backends" is out of scope for this PR. This can be discussed elsewhere.

Reply:
> the implementation should just be composable rather than ignorable parameters.

I'd prefer the latter. This is discussible.